react-components/src/components/NodeCacheProvider/PointCloudAnnotationCache.ts

```diff
@@ -0,0 +1,158 @@
+
+    return filteredAnnotationModels;
+  }
+
+  private async fetchAndCacheAssetMappingsForModel(
+    modelId: ModelId,
+    revisionId: RevisionId
+  ): Promise<Array<Map<number, Asset>>> {
+    const key = modelRevisionToKey(modelId, revisionId);
+    const annotations = await this.fetchAnnotationForModel(modelId);
+    const annotationAssets = this.fetchPointCloudAnnotationAssets(annotations, this._sdk);
+
+    this._modelToAnnotationAssetMappings.set(key, annotationAssets);
+    return await annotationAssets;
+  }
+
+  private async fetchAnnotationForModel(modelId: ModelId): Promise<AnnotationModel[]> {
+    const filter: AnnotationFilterProps = {
+      annotatedResourceIds: [{ id: modelId }],
+      annotatedResourceType: 'threedmodel',
+      annotationType: 'pointcloud.BoundingVolume'
+    };
+    return await this._sdk.annotations
+      .list({
+        filter,
+        limit: 1000
+      })
+      .autoPagingToArray({ limit: Infinity });
+  }
+
+  private async fetchPointCloudAnnotationAssets(
+    pointCloudAnnotations: AnnotationModel[],
+    sdk: CogniteClient
+  ): Promise<Array<Map<number, Asset>>> {
+    const annotationMapping = pointCloudAnnotations
+      .map((annotation) => {
+        const assetId =
+          (annotation.data as AnnotationsBoundingVolume).assetRef?.id ??
+          (annotation.data as AnnotationsBoundingVolume).assetRef?.externalId;
+        if (assetId === undefined) {
+          return undefined;
+        }
+        return {
+          annotationId: annotation.id,
+          assetId
+        };
+      })
+      .filter(
+        (mapping): mapping is { annotationId: number; assetId: number } => mapping !== undefined
+      );
+
+    const uniqueAnnotationMapping = uniqBy(annotationMapping, 'assetId');
+
+    const assetsResult = await Promise.all(
+      chunk(uniqueAnnotationMapping, 1000).map(async (uniqueMappingChunk) => {
+        const retrievedAssets = await sdk.assets.retrieve(
+          uniqueMappingChunk.map((mapping) => {
+            if (typeof mapping?.assetId === 'number') {
+              return { id: mapping.assetId };
+            } else {
+              return { externalId: mapping?.assetId };
+            }
+          }),
+          { ignoreUnknownIds: true }
+        );
+        return retrievedAssets;
+      })
+    );
+
+    const assets = assetsResult.flat();
+    const annotationToAssetMapping = assets.map((asset) => {
+      const annotationIdToAssetMap = new Map<number, Asset>();
+      uniqueAnnotationMapping.forEach((mapping) => {
+        if (mapping.assetId === asset.id) {
+          annotationIdToAssetMap.set(mapping.annotationId, asset);
+        }
+      });
+      return annotationIdToAssetMap;
+    });
+    return annotationToAssetMapping;
+  }
+
+  public async matchPointCloudAnnotationsForModel(
+    modelId: ModelId,
+    revisionId: RevisionId,
+    assetId: string | number
+  ): Promise<Array<Map<number, Asset>>> {
```

Comment:
I don't understand what this return type is for - is it really supposed to be an array of maps?

Reply:
Yes, it would contain `annotationId` & it's `Asset`

Reply:
But wouldn't all the asset objects be the same? And it wasn't obvious that the `number` here is an annotation ID :sweat_smile: . And also - why is it an array too?